src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -479,51 +536,73 @@ private NamedArgsAndClosure(Map<?,?> namedArgs, Closure body) {
      * Note that {@link DescribableModel#instantiate} would also handle {@link GString} in {@code coerce},
      * but better to do it here in the Groovy-specific code so we do not need to rely on that.
+     * Record all instances of interpolated Groovy strings. We can check this collection later to see if sensitive variables were used.
      * @return {@code v} or an equivalent with all {@link GString}s flattened, including in nested {@link List}s or {@link Map}s
      */
-    private static Object flattenGString(Object v) {
+    private static Object flattenGString(Object v, @Nonnull Set<String> interpolatedStrings) {
         if (v instanceof GString) {
-            return v.toString();
+            String flattened = v.toString();
+            interpolatedStrings.add(flattened);
+            return flattened;
         } else if (v instanceof List) {
             boolean mutated = false;
             List<Object> r = new ArrayList<>();
             for (Object o : ((List<?>) v)) {
-                Object o2 = flattenGString(o);
+                Object o2 = flattenGString(o, interpolatedStrings);
                 mutated |= o != o2;
                 r.add(o2);
             }
             return mutated ? r : v;
         } else if (v instanceof Map) {
             boolean mutated = false;
-            Map<Object,Object> r = new LinkedHashMap<>(preallocatedHashmapCapacity(((Map) v).size()));
-            for (Map.Entry<?,?> e : ((Map<?, ?>) v).entrySet()) {
+            Map<Object, Object> r = new LinkedHashMap<>(preallocatedHashmapCapacity(((Map) v).size()));
+            for (Map.Entry<?, ?> e : ((Map<?, ?>) v).entrySet()) {
                 Object k = e.getKey();
-                Object k2 = flattenGString(k);
+                Object k2 = flattenGString(k, interpolatedStrings);
                 Object o = e.getValue();
-                Object o2 = flattenGString(o);
+                Object o2 = flattenGString(o, interpolatedStrings);
                 mutated |= k != k2 || o != o2;
                 r.put(k2, o2);
             }
             return mutated ? r : v;
+        } else if (v instanceof NamedArgsAndClosure) {
+            UninstantiatedDescribable ud = ((NamedArgsAndClosure) v).uninstantiatedDescribable;
+            return ud != null? ud : v;
         } else {
             return v;
         }
     }
 
     static NamedArgsAndClosure parseArgs(Object arg, StepDescriptor d) {
+        Set<String> interpolatedStrings = new HashSet<>();
+        if (arg instanceof NamedArgsAndClosure) {
+            interpolatedStrings = ((NamedArgsAndClosure) arg).interpolatedStrings;
+        } else if (arg instanceof Object[]) {
+            Object[] array = (Object[]) arg;
+            for (Object o : array) {
+                if (o instanceof NamedArgsAndClosure) {
+                    interpolatedStrings.addAll(((NamedArgsAndClosure) o).interpolatedStrings);
+                }
+            }
+        }
```

Comment:
Looks like it can't be removed there because of how `parseArgs` handles `Object` arrays, but I was able to move this code into that specific section of `Object[]` handling .

Reply:
moved all this code to `NamedArgsAndClosure` in a8df396d68dc4f00386db06297a6ec44f923723b